### PR TITLE
fix: position indicator at start of line

### LINKED ISSUE #118: False positive matches in uppercase letter detection
The new uppercase letter detection produced some false positives for
- character classes like `\S`
- comments

### PR DESCRIPTION
**Pull request by fzipi**
Fixes #118 and #119.

- updated comment regex to match [comments that start with space](https://github.com/coreruleset/coreruleset/blob/6f8675f17f31ebebff664bbee84578f8d0a65703/regex-assembly/941210.ra#L220).
- do not match on character classes, unless we have an odd number of escapes (e.g. `\S` vs. `\\S`, the former is a char class, the latter is an escaped quote followed by uppercase `S`).
- parses definitions for uppercase also.
- added tests for everything.

### CODE REVIEW

File: cmd/regex_format.go [hunk 1]
```diff
@@ -326,16 +326,48 @@ func findUpperCaseOnIgnoreCaseFlag(lines []string, iFlag bool) (bool, string) {
 	// check if the file contains uppercase letters
 	if iFlag {
 		for number, line := range lines {
-			// ignore line if it starts with ##!
-			if strings.HasPrefix(line, "##!") {
+			// ignore line if it starts with the following regex: \s+##!
+			if regex.CommentRegex.MatchString(line) {
```
Comment (theseion) [inline cmd/regex_format.go:330]: Definition directives should not be ignored: `##!> define something "AbC-`.

File: cmd/regex_format.go [hunk 2]
```diff
@@ -326,16 +326,48 @@ func findUpperCaseOnIgnoreCaseFlag(lines []string, iFlag bool) (bool, string) {
...
 				continue
 			}
-			found := strings.IndexAny(line, "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+			found := findUppercaseNonEscaped(line)
```
Comment (theseion) [inline cmd/regex_format.go:333]: Return `(bool, int)`.

File: cmd/regex_format.go [hunk 3]
```diff
@@ -326,16 +326,48 @@ func findUpperCaseOnIgnoreCaseFlag(lines []string, iFlag bool) (bool, string) {
...
 			if found > -1 {
+				// show the column where the uppercase letter was found
+				// for a better visual match, we add equal symbols a and a caret in a line below
+				fill := ""
+				if found > 0 {
+					fill = strings.Repeat("=", found-1)
+				}
 				res = true
-				message = fmt.Sprintf("%s\n%s^ [HERE]\n", lines[number], strings.Repeat("=", found-1))
+				message = fmt.Sprintf("\n%s\n%s^ [HERE]\n", lines[number], fill)
```
Comment (theseion) [inline cmd/regex_format.go:342]:
```suggestion
				message = fmt.Sprintf("\n%s\n%s^ [HERE]\n", line, fill)
```

File: cmd/regex_format.go [hunk 4]
```diff
@@ -326,16 +326,48 @@ func findUpperCaseOnIgnoreCaseFlag(lines []string, iFlag bool) (bool, string) {
...
 				continue
 			}
-			found := strings.IndexAny(line, "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+			found := findUppercaseNonEscaped(line)
 			if found > -1 {
+				// show the column where the uppercase letter was found
+				// for a better visual match, we add equal symbols a and a caret in a line below
+				fill := ""
+				if found > 0 {
+					fill = strings.Repeat("=", found-1)
```
Comment (theseion) [inline cmd/regex_format.go:339]: Index is off by one if found > 0.
```suggestion
					fill = strings.Repeat("=", found)
```

File: cmd/regex_format.go [hunk 5]
```diff
@@ -326,16 +326,48 @@ func findUpperCaseOnIgnoreCaseFlag(lines []string, iFlag bool) (bool, string) {
...
 		for number, line := range lines {
-			// ignore line if it starts with ##!
-			if strings.HasPrefix(line, "##!") {
+			// ignore line if it starts with the following regex: \s+##!
+			if regex.CommentRegex.MatchString(line) {
 				continue
 			}
-			found := strings.IndexAny(line, "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+			found := findUppercaseNonEscaped(line)
 			if found > -1 {
```
Comment (theseion) [inline cmd/regex_format.go:334]: Check the boolean here.
```suggestion
			if found {
```

File: cmd/regex_format.go (hunk 3 above)
Comment (theseion) [inline cmd/regex_format.go:342]: Very clever to simply break to the next line! 👏

**Review by theseion — CHANGES_REQUESTED** [review]

File: cmd/regex_format.go (hunk 1 above)
Comment (fzipi) [reply cmd/regex_format.go:330]: Ugh, parsing definitions could be a rabbit hole. Let me see if there is a smarter way 🤔

File: cmd/regex_format.go (hunk 4 above)
Comment (fzipi) [reply cmd/regex_format.go:339]: No, because the last char is the caret "pointing" up.

Comment (theseion) [reply cmd/regex_format.go]: But it is, check your output :) With `0`, `fill` will be empty. With `1`, `fill` will be `=`, which means that _position_ `0` will have `=`. So the "-1" is already implied by the logic of creating the fill string.

Comment (theseion) [reply cmd/regex_format.go]: Sorry, didn't realised you had changed it.

File: cmd/regex_format.go (hunk 1 above)
Comment (fzipi) [reply cmd/regex_format.go]: Found a solution, but I would say that it feel we are repeating things from the parser here... 🤷

Comment (fzipi) [reply cmd/regex_format.go]: Ok, updating the global definition regex looks like it makes it better now.

**Review by theseion — APPROVED** [review]
Great, thanks!

**Pull request merged**